File: javascript/src/blackjack.js
// Bring in super awesome libraries for cool programming
var lodash = require('lodash');
var rls = require('readline-sync');

// Card suits
var suits = [
    'HEARTS',
    'DIAMONDS',
    'SPADES',
    'CLUBS'
];

// Card pips
var pips = [
    'ACE',
    'TWO',
    'THREE',
    'FOUR',
    'FIVE',
    'SIX',
    'SEVEN',
    'EIGHT',
    'NINE',
    'TEN',
    'JACK',
    'QUEEN',
    'KING'
];

/**
 * Class defining playing cards with pip and suit represented
 * 
 */
class Card {

    /**
     * Initializes a Card
     * 
     * @param  {String} pip
     * @param  {String} suit
     */
    constructor(pip, suit) {
        this.pip = pip,
        this.suit = suit
    }

    /**
     * Returns the printed representation of a Card
     * 
     * @return {String} Printed Representation
     */
    toString() { return this.pip + ' OF ' + this.suit; }

}

/**
 * Class defining a deck of playing cards that can be shuffled or
 * have cards dealt from the top
 * 
 */
class Deck {

    /**
     * Initializes a Deck of cards, iterating through all pips and suits
     * to generate a full deck.
     * 
     */
    constructor() {
        var suit, pip;

        this.cards = [];

        suits.map(suit => {
            pips.map(pip => {
                this.cards.push(new Card(pip, suit));
            })
        });
    }

    /**
     * Shuffles the deck by randomizing the order of the cards within
     * 
     */
    shuffle() { this.cards = lodash.shuffle(this.cards); }

    /**
     * Removes and returns the card at the top (end of the array) of the deck
     * 
     * @return {Card} The "top" card off the deck
     */
    deal() { return this.cards.pop(); }

}

/**
 * Class defining a player's hand in blackjack.
 * Can add cards, empty, print held cards, and determine score.
 * 
 */
class Hand {

    /**
     * Initializes Hand object with empty card array
     */
    constructor() { this.cards = [] }

    /**
     * Adds a card to the hand
     * 
     * @param {Card} card
     */
    add(card) { this.cards.push(card); }

    /**
     * Empties the hand
     * 
     */
    empty() { this.cards = []; }

    /**
     * Prints the hand to the console with cards numbered starting
     * at 1 (+1 to index)
     * 
     */
    print() {
        this.cards.forEach(function(card, index){
            var displayIndex = index + 1;
            console.log(displayIndex + ') ' + card.toString());
        });
    }

    /**
     * Checks whether hand score exceeds limit of 21
     * 
     * @return {Boolean} is the hand busted?
     */
    isBusted(){ return this.getScore() > 21; }

    /**
     * Calculates the hand score, saving Aces for last because they
     * are 11 if they don't bust and 1 if they do.
     * 
     * @return {Integer} score
     */
    getScore(){
        var idx, card, cardValue, ace;
        var score = 0;
        var aceCount = 0;

        // Iterate through hand, scoring cards by ordered index in `pips` array
        for (idx in this.cards) {
            card = this.cards[idx];
            if (card.pip != 'ACE') {
                cardValue = lodash.findIndex(pips, pip => pip == card.pip ) + 1;
                score = score + cardValue;
            } else {
                aceCount = aceCount + 1;
            }
        }

        // For every ace, add the relevant score
        for (ace in aceCount) {
            if ((score + 11) > 21) {
                score = score + 1;
            } else {
                score = score + 11;
            }
        }

        return score;
    }

}

/**
 * Given a score, determines the result of a round.
 * 
 * @param  {Integer} score
 */
function declareResult(score) {
    console.log('* * * * * * * * * * * * * * * *');
    if (score < 21) {
        console.log('Your score is:', score);
    } else if (score == 21) {
        console.log('Blackjack! Damn fine job.');
    } else {
        console.log('Busted! Better luck next time.');
    }
    console.log('* * * * * * * * * * * * * * * *');
}

/**
 *  Plays a round of Blackjack
 * 
 */
function playRound() {
    var deck = new Deck();
    var hand = new Hand();
    var x, option, busted;

    deck.shuffle();

    // Deal 2 cards to hand
    for (x in lodash.range(2)) {
        hand.add(deck.deal());
    }

    // Label the loop for continues/breaks
    round:
    while(true) {
        console.log('Your Hand:');
        hand.print();
        console.log('Current Score:', hand.getScore());
        if (hand.isBusted()) {
            declareResult(hand.getScore());
            break round;
        } else if (hand.getScore() == 21) {
            declareResult(hand.getScore());
            break round;
        } else {
            option = rls.question('Hit (h) or Stay (s)?\n');
        }

        switch(option.toLowerCase()) {
            // Player decides to `hit`
            case 'h':
                hand.add(deck.deal());
                continue round;
            // Player decides to `stay`
            case 's':
                declareResult(hand.getScore());
                break round;
            // Player can't read or type correctly, keep prodding
            default:
                continue round;
        }
    }

    return rls.question('Would you like to play again? (y/n)\n').toLowerCase() == 'y';
}

/**
 * Main function, holds game loop
 * 
 */
function main() {
    while(playRound()){}
}

// Run main function
main();

File: python/blackjack.py
from numpy import random as rd

suits = [
    'HEARTS',
    'DIAMONDS',
    'SPADES',
    'CLUBS'
]

pips = [
    'ACE',
    'TWO',
    'THREE',
    'FOUR',
    'FIVE',
    'SIX',
    'SEVEN',
    'EIGHT',
    'NINE',
    'TEN',
    'JACK',
    'QUEEN',
    'KING'
]


class Card():
    """
    Defines playing cards w/ pip and suit
    """

    def __init__(self, pip, suit):
        self.pip = pip
        self.suit = suit

    def __str__(self):
        """
        Produces a string representation of the card
        """
        return self.pip + ' OF ' + self.suit

    def __repr__(self):
        """
        When print()ing the card directly, uses the returned value
        """
        return self.__str__()


class Deck():
    """
    Defines a deck of playing cards
    """

    def __init__(self):
        """
        Initializes the deck with every combination of pip and suit
        """
        self.cards = []
        for suit in suits:
            for card in [Card(pip, suit) for pip in pips]:
                self.cards.append(card)

    def shuffle(self):
        """
        Randomizes order of cards in deck
        """
        rd.shuffle(self.cards)

    def deal(self):
        """
        `Draws` the top card off the deck
        """
        return self.cards.pop()


class Hand():
    """
    Defines a player's hand in blackjack
    """

    def __init__(self):
        self.cards = []

    def __str__(self):
        """
        Produces a string representation of a hand
        """
        self_string = ''

        for idx, card in enumerate(self.cards):
            self_string += str(idx + 1) + ') ' + str(card)
            if idx != (len(self.cards) - 1):
                self_string += '\n'
        return self_string

    def __repr__(self):
        """
        Allows print() calls to directly print the hand
        """
        return self.__str__()

    @property
    def score(self):
        """
        Generates `score` property on calls to Hand.score
        """
        # Generate a dictionary of pip scores
        score_dict = {pip: (idx + 1) for idx, pip in enumerate(pips)}
        score = 0
        ace_count = 0

        # Iterate through hand, scoring cards via the dictionary
        for card in self.cards:
            if card.pip != 'ACE':
                score = score + score_dict[card.pip]
            else:
                ace_count = ace_count + 1

        # For every ace, determine if high or low
        for ace in range(ace_count):
            if (score + 11) > 21:
                score = score + 1
            else:
                score = score + 11

        return score

    def add(self, card):
        """
        Adds a `drawn` card to the hand
        """
        self.cards.append(card)

    def empty(self):
        """
        Empties the hand
        """
        self.cards.clear()

    def isBusted(self):
        """
        Checks whether the hand score exceeds 21
        """
        return self.score > 21


def declare_result(score):
    """
    Given a score, determines the result of a round and prints it
    """
    print('* * * * * * * * * * * * * * * *')
    if score < 21:
        print('Your score is: ' + str(score))
    elif score == 21:
        print('Blackjack! Damn fine job.')
    else:
        print('Busted! Better luck next time.')
    print('* * * * * * * * * * * * * * * *')


def play_round():
    """
    Plays a round of Blackjack
    """
    deck = Deck()
    hand = Hand()

    deck.shuffle()

    # deal 2 cards to hand
    for x in range(2):
        hand.add(deck.deal())

    # Keep variable designating continuation of round
    round_going = True
    while round_going:
        print('Your Hand:')
        print(hand)
        print('Current Score: ' + str(hand.score))

        if hand.isBusted():
            declare_result(hand.score)
            round_going = False
            continue
        elif hand.score == 21:
            declare_result(hand.score)
            round_going = False
            continue
        else:
            option = input('Hit (h) or Stay (s)?\n')

        if option.lower() == 'h':
            # player decides to `hit`
            hand.add(deck.deal())
            continue
        elif option.lower() == 's':
            # player decides to `stay
            declare_result(hand.score)
            break
        else:
            # player can't read or type correctly, keep prodding
            continue

    return input('Would you like to play again? (y/n)\n').lower() == 'y'


# Main function, with game loop
if __name__ == '__main__':
    while play_round():
        pass
